File: src/core/analysis/storage.ts
// Water storage near the start (`water.storage_possible`, D111: it replaces `water.reservoir`; the
// workshop study's W1 rule, investigation/WORKSHOP-INTEGRATION.md §2). Information the generator
// prefers, never a guard (D209, decisions-pending #67): engineering water is the player's job. It
// asks that the land lets the player store a drought's water near the start, not that the map holds
// its water (D152):
// 1. running water: the clean water the start's pump shore touches is part of a body fed by clean
//    sources of at least need ÷ (2 × 460) blocks per second, so it refills the colony's drought need
//    in two game days;
// 2. storage is possible within 40 tiles of the start by any one of: a straight dam whose reservoir
//    holds the need (the validator's dam sampling); natural water kept through the drought; or
//    levees: water within 40 tiles of the start, raised by a crest of 1–3 levels (never above the
//    start's own level), floods ground within 60 tiles of the start, off the map's edge, that holds
//    the need, with the levee line (where the fill would run past the 60-tile box or onto the edge
//    ring) at most a quarter of the flooded area's perimeter.
// prototype/playability.py `storage_possible` is the same rule; the oracle compares them. Ported from
// the design prototype (investigation/generative/proto/storage.ts).

import type { Emitter } from "../sim/water";

/** A game day in seconds of water flow (the simulator's day). */
export const SECONDS_PER_DAY = 460;

/** Clean strength feeding the water body that `tile` belongs to (4-connected, any depth). */
export function runningFlow(W: number, H: number, depth: ArrayLike<number>, emitters: readonly Emitter[], tile: number): number {
  const N = W * H;
  const body = new Uint8Array(N);
  const q = [tile];
  body[tile] = 1;
  for (let k = 0; k < q.length; k++) {
    const i = q[k];
    const x = i % W;
    const y = (i - x) / W;
    const nb = [x > 0 ? i - 1 : -1, x + 1 < W ? i + 1 : -1, y > 0 ? i - W : -1, y + 1 < H ? i + W : -1];
    for (const j of nb) {
      if (j < 0 || body[j] || !(depth[j] > 0.001)) continue;
      body[j] = 1;
      q.push(j);
    }
  }
  let running = 0;
  for (const e of emitters) {
    if (e.contamination > 0 || !(e.strength > 0)) continue;
    if (e.cells.some((c) => body[c])) running += e.strength;
  }
  return running;
}

/** The most a dam and levees can hold near the start (rule 2's levees), up to `enough`. */
export function leveeStorage(h: ArrayLike<number>, W: number, H: number, D: ArrayLike<number>, C: ArrayLike<number>, start: { x: number; y: number; z: number }, enough: number): number {
  const N = W * H;
  let best = 0;
  const inBox = (x: number, y: number) => Math.max(Math.abs(x - start.x), Math.abs(y - start.y)) <= 60 && x > 0 && y > 0 && x < W - 1 && y < H - 1;
  const R = new Int32Array(N).fill(-1);
  let stamp = 0;
  const seeds: number[] = [];
  for (let y = Math.max(1, start.y - 40); y <= Math.min(H - 2, start.y + 40); y += 3)
    for (let x = Math.max(1, start.x - 40); x <= Math.min(W - 2, start.x + 40); x += 3) {
      const i = y * W + x;
      if (D[i] > 0.05 && C[i] < 0.05) seeds.push(i);
    }
  const tried = new Set<string>();
  for (const s0 of seeds) {
    for (let crest = 1; crest <= 3; crest++) {
      const lambda = h[s0] + crest;
      if (lambda > start.z) break;
      stamp++;
      if (R[s0] >= 0 && tried.has(`${R[s0]}:${lambda}`)) continue;
      const q = [s0];
      R[s0] = stamp;
      let cut = 0;
      let perim = 0;
      let vol = 0;
      for (let k = 0; k < q.length; k++) {
        const i = q[k];
        const x = i % W;
        const y = (i - x) / W;
        const sfc = h[i] + D[i];
        if (lambda > sfc) vol += lambda - sfc;
        for (const [dx, dy] of [[1, 0], [-1, 0], [0, 1], [0, -1]] as const) {
          const xx = x + dx;
          const yy = y + dy;
          if (xx < 0 || yy < 0 || xx >= W || yy >= H) continue;
          const j = yy * W + xx;
          if (R[j] === stamp) continue;
          if (h[j] >= lambda) {
            perim++;
            continue;
          }
          if (!inBox(xx, yy)) {
            cut++;
            perim++;
            continue;
          }
          R[j] = stamp;
          q.push(j);
        }
      }
      tried.add(`${stamp}:${lambda}`);
      if (cut <= 0.25 * perim && vol > best) best = vol;
      if (best >= enough) return best;
    }
  }
  return best;
}

File: prototype/storage.py
"""Water storage near the start and dam walls (M9a): the Python twins of
src/core/analysis/storage.ts (`water.storage_possible`, D111, replacing `water.reservoir`;
information the generator prefers, D209 #67) and src/core/analysis/ridge.ts (`terrain.dam_wall`,
D111: no built dam walls). Same rules, same order, so tools/oracle.ts compares the verdicts.

JavaScript's Math.round rounds halves up; Python's round() rounds them to even, so `_js_round` is
used wherever the TypeScript rounds."""
from __future__ import annotations

import math

SECONDS_PER_DAY = 460


def _js_round(v: float) -> int:
    return math.floor(v + 0.5)


def pump_shore_tile(walk, h, D, C, depth_min=0.3, clean_max=0.05, reach=2):
    """src/core/analysis/walk.ts pumpShoreDistance: the walk to the nearest pump shore and the water
    tile it draws from, (inf, None) when none."""
    import numpy as np
    Y, X = h.shape
    best = float("inf")
    tile = None
    for y, x in zip(*np.nonzero((D >= depth_min) & (C < clean_max))):
        surface = float(h[y, x]) + float(D[y, x])
        for dy, dx in ((0, -1), (0, 1), (-1, 0), (1, 0)):
            yy, xx = y + dy, x + dx
            if not (0 <= yy < Y and 0 <= xx < X) or not walk[yy, xx] < best:
                continue
            level = int(h[yy, xx])
            if level - reach <= surface <= level + 0.01:
                best = float(walk[yy, xx])
                tile = (int(y), int(x))
    return best, tile


def running_flow(D, sources, tile) -> float:
    """Clean strength feeding the water body `tile` (y, x) belongs to (4-connected, any depth)."""
    Y, X = D.shape
    body = set([tile])
    q = [tile]
    k = 0
    while k < len(q):
        y, x = q[k]
        k += 1
        for yy, xx in ((y, x - 1), (y, x + 1), (y - 1, x), (y + 1, x)):
            if 0 <= yy < Y and 0 <= xx < X and (yy, xx) not in body and D[yy, xx] > 0.001:
                body.add((yy, xx))
                q.append((yy, xx))
    running = 0.0
    for s in sources:
        if s["contamination"] > 0 or not s["strength"] > 0:
            continue
        if any((ty, tx) in body for ty, tx in s["tiles"]):
            running += s["strength"]
    return running


def levee_storage(h, D, C, sx, sy, sz, enough) -> float:
    """analysis/storage.ts leveeStorage: the most a dam and levees can hold near the start."""
    Y, X = h.shape
    W, H = X, Y
    hl = h.tolist()
    Dl = D.tolist()
    Cl = C.tolist()
    best = 0.0
    R = [[-1] * W for _ in range(H)]
    stamp = 0
    seeds = []
    for y in range(max(1, sy - 40), min(H - 2, sy + 40) + 1, 3):
        for x in range(max(1, sx - 40), min(W - 2, sx + 40) + 1, 3):
            if Dl[y][x] > 0.05 and Cl[y][x] < 0.05:
                seeds.append((y, x))
    tried = set()
    for (y0, x0) in seeds:
        for crest in (1, 2, 3):
            lam = hl[y0][x0] + crest
            if lam > sz:
                break
            stamp += 1
            if R[y0][x0] >= 0 and (R[y0][x0], lam) in tried:
                continue
            q = [(y0, x0)]
            R[y0][x0] = stamp
            cut = 0
            perim = 0
            vol = 0.0
            k = 0
            while k < len(q):
                y, x = q[k]
                k += 1
                sfc = hl[y][x] + Dl[y][x]
                if lam > sfc:
                    vol += lam - sfc
                for dx, dy in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                    xx, yy = x + dx, y + dy
                    if xx < 0 or yy < 0 or xx >= W or yy >= H:
                        continue
                    if R[yy][xx] == stamp:
                        continue
                    if hl[yy][xx] >= lam:
                        perim += 1
                        continue
                    if not (max(abs(xx - sx), abs(yy - sy)) <= 60 and 0 < xx < W - 1 and 0 < yy < H - 1):
                        cut += 1
                        perim += 1
                        continue
                    R[yy][xx] = stamp
                    q.append((yy, xx))
            tried.add((stamp, lam))
            if cut <= 0.25 * perim and vol > best:
                best = vol
            if best >= enough:
                return best
    return best


# ---- dam walls (analysis/ridge.ts)

C22 = 0.9238795325112867
S22 = 0.3826834323650898
R2 = 0.7071067811865476
LINES = ((1, 0), (C22, S22), (R2, R2), (S22, C22), (0, 1), (-S22, C22), (-R2, R2), (-C22, S22))


def dam_walls(h, D) -> list:
    """A straight band of rock 2–6 levels high across a valley, with vertical faces, a flat crest,
    even thickness and dry floor on both sides, and a gap for the river (analysis/ridge.ts)."""
    Y, X = h.shape
    W, H = X, Y
    hl = h.tolist()
    Dl = D.tolist()

    def at(x, y):
        xi = _js_round(x)
        yi = _js_round(y)
        return None if xi < 0 or yi < 0 or xi >= W or yi >= H else (yi, xi)

    def probe(px, py, nx, ny):
        c = at(px, py)
        if c is None or Dl[c[0]][c[1]] > 0.05:
            return None
        crest = hl[c[0]][c[1]]
        floors = []
        dry = False
        for sgn in (1, -1):
            f = math.inf
            wet_face = False
            for m in range(4, 8):
                j = at(px + sgn * m * nx, py + sgn * m * ny)
                if j is None:
                    return None
                if hl[j[0]][j[1]] < f:
                    f = hl[j[0]][j[1]]
                if Dl[j[0]][j[1]] > 0.05:
                    wet_face = True
            if not wet_face:
                dry = True
            if crest < f + 2 or crest > f + 6:
                return None
            face = False
            for m in range(0, 6):
                if face:
                    break
                a = at(px + sgn * m * nx, py + sgn * m * ny)
                b = at(px + sgn * (m + 1) * nx, py + sgn * (m + 1) * ny)
                if a is None or b is None:
                    return None
                if hl[a[0]][a[1]] >= crest - 1 and hl[a[0]][a[1]] - hl[b[0]][b[1]] >= 2:
                    face = True
            if not face:
                return None
            floors.append(f)
        thick = 1
        for sgn in (1, -1):
            for m in range(1, 9):
                j = at(px + sgn * m * nx, py + sgn * m * ny)
                if j is None or hl[j[0]][j[1]] < crest - 1:
                    break
                thick += 1
        if thick < 2 or thick > 8:
            return None
        return (crest, floors[0], floors[1], thick, dry)

    found = []
    for cy in range(H):
        row = Dl[cy]
        for cx in range(W):
            if not row[cx] > 0.05:
                continue
            if any(abs(wx - cx) + abs(wy - cy) <= 12 for wx, wy in found):
                continue
            for tx, ty in LINES:
                nx, ny = -ty, tx
                sides = []
                for sgn in (1, -1):
                    pts = []
                    started = False
                    miss = 0
                    for k in range(1, 41):
                        p = probe(cx + sgn * k * tx, cy + sgn * k * ty, nx, ny)
                        if p is not None:
                            started = True
                            miss = 0
                            pts.append(p)
                        elif not started:
                            if k > 8:
                                break
                        else:
                            miss += 1
                            if miss > 1:
                                break
                    sides.append(pts)
                if len(sides[0]) < 6 or len(sides[1]) < 6:
                    continue
                allp = sides[0] + sides[1]
                crests = [p[0] for p in allp]
                if max(crests) - min(crests) > 1:
                    continue

                def med(v):
                    return sorted(v)[len(v) >> 1]
                fa = med([p[1] for p in allp])
                fb = med([p[2] for p in allp])
                if abs(fa - fb) > 1:
                    continue
                th = [p[3] for p in allp]
                if max(th) - min(th) > 3:
                    continue
                if sum(1 for p in allp if p[4]) < 0.75 * len(allp):
                    continue
                found.append((cx, cy))
                break
    return found
